src/tests/unit/doughs-dao.test.js

```diff
@@ -0,0 +1,85 @@
+  const getDoughsQuery = `SELECT ${getDoughsConditions} FROM DOUGHS `;
+  const getDoughsQueryWithWaterTemp = `SELECT ${getDoughsConditions} FROM DOUGHS WHERE WATER_TEMP = :waterTemp ;`;
+  const getDoughsQueryWithMultipleConditions = `SELECT ${getDoughsConditions} FROM DOUGHS WHERE WATER_TEMP = :waterTemp AND GRAMS_WATER = :gramsWater AND PROOF_TIME = :proofTime AND NAME = :name ;`;
+  const executeReturnRows = ['a', 'b', 'c'];
+  const executeReturn = { rows: executeReturnRows, metadata: 'ignore-this' };
+  beforeEach(() => {
+    connectionSpy = sinon.stub().returns(executeReturn);
+    serializeDoughsStub = sinon.stub(doughsSerializer, 'serializeDoughs');
+    getConnectionStub = sinon.stub(connectionApi, 'getConnection');
+    getConnectionStub.resolves({
+      execute: connectionSpy,
+      close: () => { },
+      commit: () => { },
+    });
+    serializeDoughsStub.returns(baseGetDoughsReturn);
+  });
+  afterEach(() => {
+    sinon.restore();
+  });
+  describe('getDoughs', () => {
+    it('returns the output from serializeDoughs', () => {
+      const result = getDoughs({});
+      return result.should.eventually.deep.equal(baseGetDoughsReturn);
+    });
+    it('extracts only the rows from the database return', async () => {
+      await getDoughs({});
+      serializeDoughsStub.getCall(0).calledWith(executeReturnRows).should.equal(true);
+    });
+    describe('when it has an invalid filter', () => {
+      it('does not parse the invalid filter', async () => {
+        await getDoughs({ abc: 'def' });
+        connectionSpy.getCall(0).calledWith(getDoughsQuery, {}).should.equal(true);
+      });
+    });
+    describe('when it has valid filters', () => {
+      it('properly parses those filters into bind paramters', async () => {
+        await getDoughs({ 'filter[waterTemp]': 90 });
+        connectionSpy.getCall(0).calledWith(getDoughsQueryWithWaterTemp, { waterTemp: 90 });
+      });
+    });
+    describe('when it gets valid and invalid filters', () => {
+      it('only parses the valid filters into bind parameters', async () => {
+        await getDoughs({ 'filter[waterTemp]': 90, abc: 'def' });
+        connectionSpy.getCall(0).calledWith(getDoughsQueryWithWaterTemp, { waterTemp: 90 });
+      });
+    });
+    describe('when it gets multiple valid filters', () => {
+      it('properly parses all of those filters', async () => {
+        await getDoughs({
+          'filter[waterTemp]': 90,
+          'filter[gramsWater]': 200,
+          'filter[proofTime]': 90,
+          'filter[name]': 'Sample dough',
+        });
+        connectionSpy.getCall(0).calledWith(getDoughsQueryWithMultipleConditions, {
+          waterTemp: 90,
+          gramsWater: 200,
+          proofTime: 90,
+          name: 'Sample dough',
+        });
```

Comment:
Your test should also cover every function (include helpers like `doughColumnAliases`, `normalizeFilterName`, etc.).

Reply:
My thought there was that those are implicitly covered in tests that assume the bind parameters will be correctly parsed, and that covering them explicitly would mean testing unexported functions that aren't part of the interface. I think that if were were to look at code coverage for these tests they would in fact be hitting all of the helper functions in the file.

Reply:
No really. The main point of writing unit tests is to make sure every function should work as expected and be tested. Some corner cases could be easily ignored if you only implicitly test them by calling parents functions/methods.

Reply:
(Sorry for the long comment)

My assumption was always that a "unit" in this case is a single module, and we're testing the interface of the unit by testing the exported functions. `doughColumnAliases` and`normalizeFilterName` are internal to the `doughs-dao` module and subject to change, so testing their specific behavior seems like it would make our tests more brittle. They're also not all that complicated--`doughColumnAliases` in particular is constant and generates the same value every time. 

The behavior we care about, in my opinion, is how the `getDoughs` function generates queries and bind parameters, and how it interacts with `serializeDoughs`. That functionality fully encompasses properly normalizing filter names, and if we introduce a bug into the filter normalization functions it should propagate into our existing tests.

I do understand the alternative position though, and at the end of the day I'm not going to die on this hill. This discussion also kind of makes me wonder if it would make sense to pull some of this logic out into a separate util module that we could then test on its own, without changing the interface of `doughs-dao`.

Reply:
I'm happy to discuss this topic deeper 😀 

Actually this is a debatable and interesting topic. You can find many threads about it like [this](https://stackoverflow.com/questions/1861338/unit-test-helper-methods) and [this](https://dzone.com/articles/unit-testing-guidelines-what-to-test-and-what-not).

Take a look at this post: http://softwaretestingfundamentals.com/unit-testing/

> Do not create test cases for everything. Instead, focus on the tests that impact the behavior of the system.

but also

> Write test cases that are independent of each other. For example, if a class depends on a database, do not write a case that interacts with the database to test the class. Instead, create an abstract interface around that database connection and implement that interface with a mock object.

To me, sometimes these principles can violate with each other. The question here is how "small" unit we wanna test. Let me rephrase my statement. I think what I meant to test "everything function" is not really write test cases for every single function, instead, we should test every function which directly or indirectly impact the result. For example, if you have a helper function which only takes care of generating logging messages or error messages and the logic is pretty simple, then I totally agree it's not necessary to test it. However, in your case here since most of these helper functions are used to generate SQL query and will be executed against the DB, I guess I lean toward to test them to make sure it won't generate any bad query (like SQL injection).

I'm open to any thoughts, feel free to add your comment or concern here as well 😄 

@Ruefa what's your thought on it?

Reply:
> However, in your case here since most of these helper functions are used to generate SQL query and will be executed against the DB, I guess I lean toward to test them to make sure it won't generate any bad query (like SQL injection).

I do agree it's really important to test SQL generation, but we do have a way of doing that by testing the values that `connection.execute()` is called with. The implementation details of how that SQL is generated aren't as important, in my opinion.

Reply:
Hmm, not sure.
To assert some functions will call `connection.execute()` with an expected query and the SQL query generator will always guarantee to generate the correct query are still slightly different to me.

Reply:
I would argue that unit tests should be used to test individual chunks of code in a vacuum. I think testing an "interface" of a class is better done by the integration tests. I think it's a trap to think that a method like `doughColumnAliases` doesn't need to be tested because you should be writing tests that catch errors now and _in the future_. I would agree though that `doughColumnNames` does not need an individual test because it literally returns constant values and doesn't really have logic like `doughColumnAliases` does. Another thing to keep in mind is that our code uses external dependencies that change on a regular basis. Lodash could decide to change how `_.toPairs()` works and we would want our tests to catch that.

Reply:
But, if we can show that our function always calls `connection.execute()` with a valid query, does it matter whether or not our sql generator generates the correct query under every circumstance? Because the sql generator is unexported, any case where a user can pass it parameters to generate invalid SQL will be a case we can replicate by testing the interface of the exported functions.

Reply:
Currently, the process of `getDoughs` is:

1. Take raw filters and normalize it.
2. Use the normalized filter to generate SQL query.
3. Call `connection.execute()` and execute with the generated SQL query.

Since the logic is pretty straightforward so far, only testing the step 3 should also cover the logic of step 1-2 to a certain degree. However, what if the process changed a little bit to something like this:

1. Take raw filters and normalize it.
2. Use the normalized filter to generate SQL query.
3. Call another helper function (called function A) to optimize the query (e.g. check if we can remove some unnecessary conditions or use `JOIN` clause to replace some subselect queries).
4. Call `connection.execute()` and execute with the optimized SQL query.

Assume there is a bug in step 1-2 and another one in step 3 (function A), but for some reasons these bugs coincidentally fix each other, what should we do? If you only test the final step, yes, the input query is what we expected, but does the testing process really cover what we should test?  Let's say if you add another function has a very similar logic in the future:

1. Take raw filters and normalize it.
2. Use the normalized filter to generate SQL query.
3. Call another helper function (called function B) to optimize the query.
4. Call `connection.execute()` and execute with the optimized SQL query.

If the function B is bug-free, theoretically the unit test will fail at step 4 since there's a bug in step 1-2, but how can you tell the bug is actually in step 1-2 instead of in the step 3 (function B)? If there is more than one step between 2-4, it'll be even harder to tell where the issue is, right?

Reply:
I am undecided on this issue myself. Generally, I think of a 'unit' as an exported module because the exported unit can be a dependency for any number of other functions, thus should be tested. With helper functions, they are only called through their parent function, thus act more as a way to simplify code rather than to act as its own 'unit'. That being said, I do see the value in writing unit tests for helper functions. As Liang and Alex have brought up, if there is a bug or change in a helper that breaks the module, an additional unit test would allow the issue to be identified and fixed much more easily. Ultimately, I think that it is worth it to write unit tests for helper functions, as long as they are non trivial.

Reply:
Personally, I don't think it makes sense to write unit tests for `doughColumnNames` or `doughColumnAliases`. These functions are essentially private methods to the dough dao class and I don't think it makes sense to test private methods in general.

When writing tests for methods in a module, there's a reason you aren't able to stub/mock other methods in that same module: you would essentially be modifying the module under test. If we were to undo the refactoring of `doughColumnAliases` and directly paste the logic in place of each call to it, there would be no functional difference in how the code works, since this function isn't being exported and used by other modules. But in this case, we obviously wouldn't want to modify those lines of code for testing, even if we could do that. Similarly, we wouldn't want to refactor arbitrary lines of code like [this](https://github.com/osu-mist/pizza-api/blob/b0b9e0753417a8751e4d1c9ace6ece02610721da/src/api/v1/db/oracledb/doughs-dao.js#L88-L92) into methods so that we could export them for testing, even if these tests might provide some value.

I think the reason for this is that if we later modify the dao module, even if the results of the exported functions are still correct, the unit tests could fail due to renaming, removing, or rewriting helper functions. We would also be obligated to add more tests for new helper functions. To me, this seems like unnecessary maintenance. I could also see this discouraging refactoring.

@tsoliangwu0130 from your previous example, I see how having unit tests on the helper functions could provide value by helping to pinpoint errors, but I don't think that's the only metric we should go off of. I think this would start overloading the purpose of unit tests from testing public method correctness -> testing public method correctness + debugging. In your example, if we are in the first process, even if the helper function is incorrect, the exported function is correct, which is what we're testing. As soon as we transition to the second process, we have an exported function that's incorrect, which would mean we have a failing unit test. I think it would be up to us at that point to debug the failing exported function.

If these kinds of helper functions start becoming large or complex, we probably want to move them to a separate module to avoid bloating the original module. At that point, we would want to add unit tests for that new module, since it's distinct from the original. But for simple helper functions, I don't think it makes sense to unit test them. Not as important, but it would also require us to add it to our exports and probably have to comment somewhere that the method is only being exported for testing to avoid confusion.

Reply:
@ianbrown9475 really well put. I would agree that the case @tsoliangwu0130 suggested would require a refactor of the test code, since it significantly increases the complexity of the module internals. However, I think that refactor could be as simple as stubbing the SQL optimizer and testing which inputs we pass to that stub while also testing that we pass the stub's return on to the execute function.

Reply:
I am wondering if code coverage metrics would help resolve this at all. It seems like one good indicator of an unexported method that needs to be tested on its own might be if it's difficult to cover most of the lines of the method by just testing the module's exported methods. Not very familiar with common javascript coverage tools though.

Reply:
Here's the output of `nyc --reporter=lcov --reporter=text gulp test`, per Ian's suggestion:
```
------------------------------------------------------|----------|----------|----------|----------|-------------------|
File                                                  |  % Stmts | % Branch |  % Funcs |  % Lines | Uncovered Line #s |
------------------------------------------------------|----------|----------|----------|----------|-------------------|
All files                                             |     76.7 |    42.11 |    54.55 |    77.58 |                   |
 pizza-api                                            |    88.46 |      100 |    57.14 |      100 |                   |
  gulpfile.js                                         |    88.46 |      100 |    57.14 |      100 |                   |
 pizza-api/dist/api/v1/db/oracledb/api/v1/db/oracledb |    80.28 |       50 |    68.75 |    80.88 |                   |
  connection.js                                       |       50 |    16.67 |        0 |       52 |... 52,54,55,57,58 |
  doughs-dao.js                                       |    97.78 |      100 |    91.67 |    97.67 |               130 |
 pizza-api/dist/api/v1/serializers/api/v1/serializers |    78.57 |      100 |        0 |    78.57 |                   |
  doughs-serializer.js                                |    78.57 |      100 |        0 |    78.57 |          22,23,31 |
 pizza-api/dist/tests/unit/tests/unit                 |      100 |      100 |      100 |      100 |                   |
  test-data.js                                        |      100 |      100 |      100 |      100 |                   |
 pizza-api/dist/utils/utils                           |    66.67 |    39.29 |    33.33 |    66.13 |                   |
  jsonapi.js                                          |       25 |        0 |        0 |       25 |... 52,60,62,73,81 |
  load-openapi.js                                     |       80 |      100 |      100 |       80 |             22,23 |
  logger.js                                           |    64.29 |        0 |        0 |    64.29 |    49,55,66,68,70 |
  paginator.js                                        |      100 |    91.67 |      100 |      100 |                13 |
  uri-builder.js                                      |    84.62 |        0 |       50 |    83.33 |             32,33 |
------------------------------------------------------|----------|----------|----------|----------|-------------------|
```

Reply:
@jemisonf using code coverage metrics sounds probably a good idea, but I'm not sure which one should we use. The one you shared looks like a popular one, but I think we still need to do some more investigation.

@ianbrown9475 hmm, shouldn't [the purpose of the unit test](http://softwaretestingfundamentals.com/unit-testing/) includes debugging? I guess to me it's not important if the current unit is whether private or exported, the main point of writing unit tests is to help us easily detect and locate the issues/bugs if there's any, and even further to allow us easily extend our current logic.

But anyway, overall that sounds fair to me and I also agreed we shouldn't unit test those are trivial. I'd suggest let's discuss this on our team meeting tomorrow and see what other's opinions are.